File: fullstack/back/src/util/snowflake.ts
/**
 * SnowFlake的结构如下(共64bits，每部分用-分开):
 *   0 - 0000000000 0000000000 0000000000 0000000000 0 - 00000 - 00000 - 000000000000
 *   |   ----------------------|----------------------   --|--   --|--   -----|------
 * 1bit标识                41bit 时间戳                  数据标识id 机器id     序列号id
*/

// 雪花算法
export default class SnowFlake {
    // private readonly twepoch = 1548988646430n
    private readonly twepoch = 1661397758514n

    private readonly workerIdBits: bigint = 5n // 标识ID
    private readonly dataCenterIdBits: bigint = 5n // 机器ID
    private readonly sequenceBits: bigint = 12n // 序列ID

    private readonly maxWorkerId: bigint = -1n ^ (-1n << this.workerIdBits)
    private readonly maxDataCenterId: bigint =
        -1n ^ (-1n << this.dataCenterIdBits)
    private readonly sequenceMask: bigint = -1n ^ (-1n << this.sequenceBits)

    private readonly workerIdShift: bigint = this.sequenceBits
    private readonly dataCenterIdShift: bigint =
        this.sequenceBits + this.workerIdBits
    private readonly timestampLeftShift: bigint =
        this.dataCenterIdShift + this.dataCenterIdBits

    private sequence: bigint = 0n
    private lastTimestamp: bigint = -1n

    private readonly workerId: bigint
    private readonly dataCenterId: bigint
    constructor(workerId: bigint, dataCenterId: bigint) {
        if (workerId > this.maxWorkerId || workerId < 0n) {
            throw new Error(
                `workerId can't be greater than ${this.maxWorkerId} or less than 0`,
            )
        }
        if (dataCenterId > this.maxDataCenterId || dataCenterId < 0n) {
            throw new Error(
                `dataCenterId can't be greater than ${this.maxDataCenterId} or less than 0`,
            )
        }
        this.workerId = workerId
        this.dataCenterId = dataCenterId
        return this
    }

    public nextId(): bigint {
        let timestamp = this.currentLinuxTime()
        const diff = timestamp - this.lastTimestamp

        if (diff < 0n) {
            throw new Error(
                `Clock moved backwards. Refusing to generate id for ${-diff} milliseconds`,
            )
        }

        if (diff === 0n) {
            this.sequence = (this.sequence + 1n) & this.sequenceMask
            if (this.sequence == 0n) {
                timestamp = this.tilNextMillis(this.lastTimestamp)
            }
        } else {
            this.sequence = 0n
        }

        this.lastTimestamp = timestamp

        return (
            ((timestamp - this.twepoch) << this.timestampLeftShift) |
            (this.dataCenterId << this.dataCenterIdShift) |
            (this.workerId << this.workerIdShift) |
            this.sequence
        )
    }

    private tilNextMillis(lastTimeStamp: bigint) {
        let timestamp: bigint = this.currentLinuxTime()
        while (timestamp <= lastTimeStamp) {
            timestamp = this.currentLinuxTime()
        }
        return timestamp
    }

    private currentLinuxTime(): bigint {
        return BigInt(new Date().valueOf())
    }
}

File: java/src/main/java/io/nigelwy/javaassignments/util/Snowflake.java
package io.nigelwy.javaassignments.util;

/**
 * SnowFlake id generator, inspired by Twitter.
 *
 * @link https://developer.twitter.com/en/docs/basics/twitter-ids
 * This class should be plain java class, and not depends on any framework or other package.
 * <p>
 * 64 bits long type in java
 * |-- 1 bit not use --|-- 59 bits timestamp in milliseconds --|-- 3 bits worker id --|-- 1 bits sequence --|
 * <p>
 * all timestamp values in this class are millisecond.
 */
public class Snowflake {

    /**
     * timestamp start from Sat Apr 23 2022 12:24:02 GMT+0800 (China Standard Time)
     * DO NOT change this value !
     */
    private static final long EPOCH = 1650687842890L;

    /**
     * worker-id take 3 bits
     */
    private static final long WORKER_ID_BITS = 3L;

    /**
     * sequence take 1 bits
     */
    private static final long SEQUENCE_BITS = 1L;

    /**
     * mask for some bit operation on sequence
     */
    private static final long SEQUENCE_MASK = (1L << SEQUENCE_BITS) - 1;

    /**
     * worker-id left-shift length
     */
    private static final long WORKER_ID_LEFT_SHIFT_BITS = SEQUENCE_BITS;

    /**
     * timestamp left-shift length
     */
    private static final long TIMESTAMP_LEFT_SHIFT_BITS = WORKER_ID_LEFT_SHIFT_BITS + WORKER_ID_BITS;
    public static final Long MIN_VALUE = (1L << TIMESTAMP_LEFT_SHIFT_BITS)
            | (1L << WORKER_ID_LEFT_SHIFT_BITS);
    /**
     * worker-id should not greater than this value
     */
    private static final long WORKER_ID_MAX_VALUE = (1L << WORKER_ID_BITS) - 1;
    /**
     * machine time may go backwards, if under 10 milliseconds we wait
     */
    private static final int MAX_TIMESTAMP_BACKWARDS_TO_WAIT = 10;
    private final long workerId;

    private long sequence = 0;

    private long lastTimestamp = -1L;

    /**
     * @param workerId 12-bits number
     */
    public Snowflake(long workerId) {
        if (workerId > WORKER_ID_MAX_VALUE || workerId < 1) {
            throw new IllegalArgumentException(String.format("worker id can't be greater than %d or less than 1", WORKER_ID_MAX_VALUE));
        }
        this.workerId = workerId;
    }

    public synchronized long getId() {
        long currentTimestamp = getCurrentTimestamp();
        // if clock moved back we may wait
        if (tolerateTimestampBackwardsIfNeed(currentTimestamp)) {
            currentTimestamp = getCurrentTimestamp();
        }
        // if current timestamp equals to previous one, we try to increase sequence
        if (lastTimestamp == currentTimestamp) {
            if (sequenceIncreaseIfReachLimitReset()) {
                currentTimestamp = waitUntilNextTime(currentTimestamp);
            }
        }
        // we go into to new timestamp, reset sequence
        else {
            sequence = 0L;
        }
        lastTimestamp = currentTimestamp;
        return ((currentTimestamp - EPOCH) << TIMESTAMP_LEFT_SHIFT_BITS)
                | (workerId << WORKER_ID_LEFT_SHIFT_BITS)
                | sequence;
    }

    private boolean tolerateTimestampBackwardsIfNeed(long curTimestamp) {
        if (lastTimestamp <= curTimestamp) {
            return false;
        }
        long timeDifference = lastTimestamp - curTimestamp;
        if (timeDifference < MAX_TIMESTAMP_BACKWARDS_TO_WAIT) {
            waitUntilNextTime(lastTimestamp);
        } else {
            throw new IllegalStateException("Machine time goes backwards more than 10 millis");
        }
        return true;
    }

    private boolean sequenceIncreaseIfReachLimitReset() {
        return 0L == (sequence = (sequence + 1) & SEQUENCE_MASK);
    }

    private long waitUntilNextTime(long timestampToContinue) {
        long timestamp;
        do {
            timestamp = getCurrentTimestamp();
        } while (timestamp <= timestampToContinue);
        return timestamp;
    }

    private long getCurrentTimestamp() {
        return System.currentTimeMillis();
    }
}
